core/src/consensus/tendermint/mod.rs

```diff
@@ -1197,29 +1471,50 @@ impl ConsensusEngine<CodeChainMachine> for Tendermint {
     /// This operation is synchronous and may (quite reasonably) not be available, in which case
     /// `Seal::None` will be returned.
     fn generate_seal(&self, block: &ExecutedBlock, parent: &Header) -> Seal {
-        let guard = self.inner.lock();
-        guard.generate_seal(block, parent)
+        let (result, receiver) = crossbeam::bounded(1);
+        let block_number = block.header().number();
+        let parent_hash = parent.hash();
+        self.inner
+            .send(InnerEvent::GenerateSeal {
+                block_number,
+                parent_hash,
+                result,
+            })
+            .unwrap();
+        receiver.recv().unwrap()
     }
 
     /// Called when the node is the leader and a proposal block is generated from the miner.
     /// This writes the proposal information and go to the prevote step.
     fn proposal_generated(&self, sealed_block: &SealedBlock) {
-        let mut guard = self.inner.lock();
-        guard.proposal_generated(sealed_block);
+        // TODO
```

Comment:
What is the purpose of the TODO?

Reply:
Sorry. It should be removed. I'll update the commit.